lib/report-builder-factory/report-builder-json.js

```diff
@@ -0,0 +1,225 @@
+    }
+
+    getCurrAttempt(formattedResult) {
+        const previousResult = this._getPreviousResult({formattedResult});
+
+        return shouldUpdateAttempt(previousResult.status) ? previousResult.attempt + 1 : previousResult.attempt;
+    }
+
+    _addTestResult(formattedResult, props) {
+        const testResult = this._createTestResult(formattedResult, _.extend(props, {attempt: formattedResult.attempt}));
+        const node = this._getResultNode(formattedResult);
+        const stateInBrowser = this._getStateInBrowser(node, formattedResult);
+
+        if (!stateInBrowser) {
+            this._initNode(node, formattedResult, testResult);
+            formattedResult.image = hasImage(formattedResult);
+            return formattedResult;
+        }
+
+        this._updateNode({node, stateInBrowser, formattedResult, testResult});
+        return formattedResult;
+    }
+
+
+
+    _initNode(node, formattedResult, testResult) {
+        const {browserId} = formattedResult;
+        extendTestWithImagePaths(testResult, formattedResult);
+
+        if (hasNoRefImageErrors(formattedResult)) {
+            testResult.status = FAIL;
+        }
+
+        node.browsers.push({name: browserId, result: testResult, retries: []});
+        setStatusForBranch(this._tree, node.suitePath);
+    }
+
+    _updateNode({node, stateInBrowser, formattedResult, testResult}) {
+        const previousResult = this._getPreviousResult({formattedResult, stateInBrowser});
+        const {imagesInfo} = stateInBrowser.result;
+        const newResult = extendTestWithImagePaths(testResult, formattedResult, imagesInfo);
+        const shouldUpdateResult = this._shouldUpdateResult(formattedResult, previousResult);
+        const shouldAddRetry = this._shouldAddRetry(testResult, previousResult);
+
+        if (shouldAddRetry && !shouldUpdateResult) {
+            stateInBrowser.retries[formattedResult.attempt] = newResult;
+        }
+
+        if (shouldUpdateResult) {
+            if (shouldAddRetry) {
+                stateInBrowser.retries[previousResult.attempt] = previousResult;
+            }
+            formattedResult.image = hasImage(formattedResult);
+
+            const {status: currentStatus} = stateInBrowser.result;
+            stateInBrowser.result = newResult;
+
+            this._setResultStatus(stateInBrowser, currentStatus);
+        }
+
+        setStatusForBranch(this._tree, node.suitePath);
+    }
+
+    _shouldAddRetry(testResult, previousResult) {
+        const statuses = [SKIPPED, RUNNING, IDLE];
+
+        return !statuses.includes(previousResult.status) && testResult.status !== UPDATED;
+    }
+
+    _shouldUpdateResult(formattedResult, previousResult) {
+        return formattedResult.attempt >= previousResult.attempt;
+    }
+
+    save() {
+        return this.saveDataFileAsync()
+            .then(() => this._copyToReportDir(['index.html', 'report.min.js', 'report.min.css']))
+            .then(() => this)
```

Comment:
This is code previously written by someone else  [in report-builder.js](https://github.com/gemini-testing/html-reporter/blob/0d82f702b2887b8f3c4220a0091a9495964bf265/lib/report-builder-factory/report-builder.js#L235-L240

)

Reply:
You can fix it in one second, right? )

Reply:
without .then(() => this) it doesn't return correct html report file path (it outputs 
> Your HTML report is here: file://undefined
) because it uses the returned this here:
https://github.com/gemini-testing/html-reporter/blob/0d82f702b2887b8f3c4220a0091a9495964bf265/lib/plugin-adapter.js#L56-L58